[WIP] DistributedActors and distributed funcs in protocols must survive hop_to them

ktoso opened the pull request:
Resolves rdar://84054772

Declaring a protocol that is `protocol X: DistributedActor` and having a distributed func in an extension on it, causes crashes in two situations:

- attempting to get an executor off it, triggers assertions that a non Actor type tries to get an executor -- this is fine for a distributed actor though; for now we just always treat it as Default and get the global executor; we'll align the two later
- SIL synthesis of the distributed thunk needs to find the ActorSystem of the protocol's where clause, or some type alias along the way and use it during synthesis so we don't crash not being able to find the system field (and type).

The latter part I'm still struggling with, but getting through it.

ktoso committed:
[Distributed] dist actor always has default executor (currently)

ktoso committed:
[Distributed] extra test for missing makeEncoder

ktoso committed:
[DistributedDecl] Add DistributedActorSystem to known SDK types

ktoso committed:
[DistributedActor] ok progress on getting the system via witness

xedin commented on lib/AST/DistributedDecl.cpp:
```diff
@@ -65,19 +65,83 @@ using namespace swift;
+    nominal = ext->getExtendedNominal();
+  }
+
+  if (auto p = dyn_cast_or_null<ProtocolDecl>(nominal)) {
```
All you need to do is this:

```
 if (auto *protocol = DC->getSelfProtocolDecl()) {
    GenericSignature signature;
    if (auto *genericContext = member->getAsGenericContext()) {
      signature = genericContext->getGenericSignature();
    } else {
      signature = DC->getGenericSignatureOfContext();
    }

    auto ActorSystemAssocType =
        DA->getAssociatedType(C.Id_ActorSystem)->getDeclaredInterfaceType();

    auto systemTy = signature->getConcreteType(ActorSystemAssocType);
    assert(systemTy && "couldn't find concrete type for actor system");
    return systemTy;
  }
```

ktoso committed:
[SIL] SIL block names available in NOT-NOT-DEBUG builds. #41507

ktoso committed:
[Distributed] allow hop-to `let any: any X` where X is DistActor

xedin committed:
[Distributed] AST: Add an accessor to determine whether type is distributed actor

- Classes have specialized method on their declarations
- Archetypes and existentials check their conformances for
  presence of `DistributedActor` protocol.

xedin committed:
[Distributed] AST: Account for distributed members declared in class extensions

`getConcreteReplacementForProtocolActorSystemType` should use `getSelfClassDecl`
otherwise it wouldn't be able to find actor if the member is declared in an extension.

ktoso commented:
replaced by https://github.com/apple/swift/pull/41616
